[WIP] Time estimate and formatting flexibility for the Progress Meter

jbarnoud opened the pull request:
Fixes #928 

Changes made in this Pull Request:
- `mda.lib.log.ProgressMeter` get an _elapsed_, a _time_per_iteration_, and
  an _estimate_time_to_completion_ attributes that can be used in the
  format string.
  - `mda.lib.log.ProgressMeter` get a `progress` attribute to display a visual progress bar.
## PR Checklist
- [ ] Tests?
- [ ] Docs?
- [ ] CHANGELOG updated?
- [x] Issue raised/referenced?

jbarnoud commented:
I am not sure how to test the feature without adding a 10 min loop of `time.sleep`. Any idea is welcome.

kain88-de commented on package/MDAnalysis/lib/log.py:
```diff
@@ -335,6 +361,19 @@ def update(self, step, **kwargs):
+        self._last_time = datetime.datetime.now()
+        if self._start_time is None:
+            self._start_time = self._last_time
+        self.elapsed = self._last_time - self._start_time
```
Etc time completion can be calculated in a separate function that you can test.

jbarnoud replied:
It would be cleaner indeed. I obviously need some coffee.

kain88-de commented on package/MDAnalysis/lib/log.py:
```diff
@@ -335,6 +361,19 @@ def update(self, step, **kwargs):
         self.numouts += 1
 
+        # Update time tracking and estimate time to completion
+        self._last_time = datetime.datetime.now()
```
To test this just check that the values are set when you call update.

```python
pm.upate()
# store values
pm.update()
# check start_time wasn't overwritten.
```

That should be enough to check that it functions correct.

kain88-de commented:
Nive work. Since the first estimates are only in after the second iteration how to you display the etc and time per iteration until then?

kain88-de commented on package/MDAnalysis/lib/log.py:
```diff
@@ -210,6 +210,93 @@ def _guess_string_format(template):
+        return formatted
+
+    @property
+    def in_days(self):
```
how long do your scripts run that you need an estimate in days?

jbarnoud replied:
I just killed a script that ran for two weeks. Sometime a naive implementation you are confident about is more practical that spending the same amount of time trying to optimize the code.

jbarnoud commented:
I added a `start` method to the progress meter for a better time tracking. The call to `start` is optional, but making it will allows an accurate measure of the elapsed time, and therefore a more accurate estimate time of completion.

Also, I was not satisfied at all of the formatting allowed by `timedelta` objects. It meant that `elapsed` and `time_per_iteration` could only be displayed in a singled way, which was not very appropriate. I now use a modified object that allows  more flexible formatting.

Finally, I added a `progress` attribute to display a progress bar in the progress meter.

Much more details in the commit messages.

jbarnoud commented:
A quick demp: https://gist.github.com/jbarnoud/3980aa1ebe1f44a1f3962ed2c0b58379

richardjgowers commented:
Looks very cool!

mnmelo commented:
Very cool!

I have a feature request (or maybe you already include it): can it also work GROMACS-style, that for ETCs under 5 min shows the seconds to finish instead of the formatted end date/time?

jbarnoud commented:
@mnmelo Seems tricky but doable. I'll think about it.

orbeckst commented:
The gist already shows off a number of different formats (and the one with black/shaded areas looks very professional). It would be neat (read: convenient for lazy people like me...) to have a "library" of formatting options. Might be as simple as a dict of format strings in the same module that one can use.

jbarnoud commented:
@orbeckst I had the library in mind. It will come after the documentation and the tests. If anybody want to suggest formats, I'll take it.

kain88-de commented:
@jbarnoud what is missing for this?

jbarnoud commented:
It needs some cleaning, and some documentation. I also have to finish the formating of the time delta, it is missing some basic features. There is not much work left, but I do not have much time right now.

jbarnoud commented:
After a few month of rest, I revived this PR.

* I rebased against develop
* I fixed the formatting of the time delta
* I added the progressive interval fancied by @mnmelo 
* I added some doc for the additions

I still have to make the progress bar formatting behave as documented, and to test all of this.

kain88-de commented:
@jbarnoud to revive this PR should we simplify it? Maybe instead of a library just have one option.  It would just show the elapsed time and the current performance in iterations per second.

```
 4/50 [██░░░░░░░░░   8.0%] elapsed: 00:01:00 | it/s: 2000
```

 I personally like this because my own analysis usually runs in minutes to hours only and not days (other might differ here). But it gives me good information how far I am into the analysis and what my performance is. This will also tell me if I might not want to change `step` to a higher value.  Of course having the *ETC* time can be valuable too and we maybe really want to include it.

Anyway if one option is easier to implement and maintain and getting this done I'm advocating for it.

jbarnoud commented:
On 12/03/2017 08:42 PM, Max Linke wrote:
>
> @jbarnoud <https://github.com/jbarnoud> to revive this PR should we 
> simplify it? Maybe instead of a library just have one option. It would 
> just show the elapsed time and the current performance in iterations 
> per second.
>
> |4/50 [██░░░░░░░░░ 8.0%] elapsed: 00:01:00 | it/s: 2000 |
>
> I personally like this because my own analysis usually runs in minutes 
> to hours only and not days (other might differ here). But it gives me 
> good information how far I am into the analysis and what my 
> performance is. This will also tell me if I might not want to change 
> |step| to a higher value. Of course having the /ETC/ time can be 
> valuable too and we maybe really want to include it.
>
> Anyway if one option is easier to implement and maintain and getting 
> this done I'm advocating for it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/MDAnalysis/mdanalysis/pull/960#issuecomment-348808530>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABUWuurioeLnExXSrvggA3ZX7GLWHMTXks5s8vmzgaJpZM4JvMdr>.
>
This PR only needs test (and maybe some docs too). I am a bit under the 
water until the end of the year, but I regret on a daily basis not 
having that finished as it would be useful to me often. I'll try to 
finish that soon.

I do sometimes have analyses that last for more than a day. Either 
because the trajectory is ridiculously large, or because I have a bug. 
In both case, it is convenient to be able to handle more than the hours; 
like it is convenient that gromacs can display the year in its ETC.

orbeckst commented:
@jbarnoud what is the minimum that is required to be done to get this feature in?

jbarnoud commented:
On 03/14/2018 07:42 PM, Oliver Beckstein wrote:
>
> @jbarnoud <https://github.com/jbarnoud> what is the minimum that is 
> required to be done to get this feature in?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/MDAnalysis/mdanalysis/pull/960#issuecomment-373132449>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABUWugDbQRTFqxhrBtNjGJAk57GxBaLeks5teWSTgaJpZM4JvMdr>.
>
A massive rebase and some tests.

orbeckst commented:
I think our `ProgressMeter` needs a face-lift. @jbarnoud 's new version looked very slick in beta. But given how precious developer time is, I think the reality is that not much progress will be made in the next few months (pun intended). 

I wanted to explore options for an update to ProgressMeter and if there might be a simpler upgrade path, namely using [tqdm](https://tqdm.github.io/), which looks as if it could be finagled into a drop-in replacement for the current `ProgressMeter`.

I know that many people use it in their own code as
```python
import tqdm

for ts in tqdm.tqdm(u.trajectory):
   # do stuff
```

@jbarnoud does your `ProgressMeter` provide advantages over `tqdm`, i.e., features that would make it worthwhile putting extra effort into this PR?

Any other opinions? E.g., would we be ok in adding `tqdm` or something similar as a dependency?

richardjgowers commented:
I’ve used tqdm on another project and I like it. There’s a different version/callable for notebooks iirc, but maybe we can write some code which detects the notebook environment and adapts?
  

  
  

  
  
>   
> On Apr 19, 2019 at 18:22,  <Oliver Beckstein (mailto:notifications@github.com)>  wrote:
>   
>   
>   
>
> I think our  ProgressMeter  needs a face-lift.  @jbarnoud (https://github.com/jbarnoud)  's new version looked very slick in beta. But given how precious developer time is, I think the reality is that not much progress will be made in the next few months (pun intended).
>
>   
>
> I wanted to explore options for an update to ProgressMeter and if there might be a simpler upgrade path, namely using  tqdm (https://tqdm.github.io/), which looks as if it could be finagled into a drop-in replacement for the current  ProgressMeter.
>
>   
>
> I know that many people use it in their own code as
>
>   
>   
>   import  tqdm  for  ts  in  tqdm.tqdm(u.trajectory):  #  do stuff   
>
>   
>
> @jbarnoud (https://github.com/jbarnoud)  does your  ProgressMeter  provide advantages over  tqdm, i.e., features that would make it worthwhile putting extra effort into this PR?
>
>   
>
> Any other opinions? E.g., would we be ok in adding  tqdm  or something similar as a dependency?
>
>   
>
> —
>  You are receiving this because you commented.
>  Reply to this email directly,   view it on GitHub (https://github.com/MDAnalysis/mdanalysis/pull/960#issuecomment-484962749), or   mute the thread (https://github.com/notifications/unsubscribe-auth/ACGSGB67ON7FFJO4CZU7PV3PRH5UDANCNFSM4CN4Y5VQ).
>
>   
>

kain88-de commented:
Short info about tqdm usage with notebooks and jupyter lab. I've used the library extensively last year. To get the jupyterlab interface working you have to install a bunch of extra js libraries outside of pip. This was still buggy so I stuck with the normal interface. This usually works fine on jupyter lab and notebook. However, if I need to interrupt execution once the printing of tqdm is somehow broken. Instead of replacing a line it will always print a new line. This lead to long output for some cells that should only show some text with a progress meter.

I really like the library just be aware of those issues.

jbarnoud commented:
I have been using TQDM as well recently, the more I use it, the less it makes sense to me to reinvent the wheel here. TQDM is a fairly light dependency, from what I saw it is both pip installable and available on conda-forge.

orbeckst commented:
This sounds to me that we should close this PR and update #928 with the suggestion to use `tqdm`. We can have a discussion at #928 how to integrate it. tqdm's usage is different from the current `pm` – there's no explicit call needed.
